### DISCUSSION TITLE
Brotli support

### DISCUSSION THREAD

**Discussion by PhilippMolitor** [mathieucarbou/ESPAsyncWebServer#42] [Ideas]
In `WebHandlers.cpp`, the method `bool AsyncStaticWebHandler::_fileExists` implements looking for gzip files.
I would love to see it extended to also look for brotli files, as this is the emerging future standard, which also is more efficient than gzip - that is very good for use in embedded devices of course.

Also, I cannot see the `AsyncStaticWebHandler` setting a content type depending on the original file extension - is this something to improve as well?

**Comment by mathieucarbou** [comment]
Brotli support can be done with your own handlers. See examples in the ESP-dASH: https://github.com/ayushsharma82/WebSerial/blob/master/src/WebSerial.cpp

> as this is the emerging future standard

This is implemented since years already. Sadly it only available for https. You can read the reason on a discussion in the Chromium group - mainly due to how proxy work and re-gzip data and overwrite the content encoding, losing the information that it was brotli-encoded.

>  that is very good for use in embedded devices of course.

Brotli is not available on Chromium (Edge,  Chrome, etc) and Firefox over http. Only Safari allows it. So it is only OK to use it if your device uses https (and ESPAsyncWebServer cannot do https on ESP32). So what's the point...

> Also, I cannot see the `AsyncStaticWebHandler` setting a content type depending on the original file extension - is this something to improve as well?

It is because it is used in cases for static content like images, sounds, scripts, etc and the content type is known from the browser or can be set by the browser. These handlers are not to be used for REST API or dynamically consumed data.

**Comment by vortigont** [comment]
I would add to this that you can use [zopfli](https://github.com/google/zopfli) to compress you files to gz, it gives a bit better compression than native gzip.
IMHO google spoiled the whole idea of brotli with intentionally ignoring 'content-encoding' header in server's reply over plain HTTP. They are violating standards same way as some ugly proxies does trying to decompress something that they do not understand.
